shared/js/content-scope/utils.js

```diff
@@ -0,0 +1,111 @@
+import sjcl from './sjcl'
+
+export function getDataKeySync (sessionKey, domainKey, inputData) {
+    // eslint-disable-next-line new-cap
+    const hmac = new sjcl.misc.hmac(sjcl.codec.utf8String.toBits(sessionKey + domainKey), sjcl.hash.sha256)
+    return sjcl.codec.hex.fromBits(hmac.encrypt(inputData))
+}
+
+// linear feedback shift register to find a random approximation
+export function nextRandom (v) {
+    return Math.abs((v >> 1) | (((v << 62) ^ (v << 61)) & (~(~0 << 63) << 62)))
+}
+
+const exemptionLists = {}
+export function shouldExemptUrl (type, url) {
+    for (const regex of exemptionLists[type]) {
+        if (regex.test(url)) {
+            return true
+        }
+    }
+    return false
+}
+
+export function initStringExemptionLists (args) {
+    const { stringExemptionLists } = args
+    for (const type in stringExemptionLists) {
+        exemptionLists[type] = []
+        for (const stringExemption of stringExemptionLists[type]) {
+            exemptionLists[type].push(new RegExp(stringExemption))
+        }
+    }
+}
+
+// Checks the stack trace if there are known libraries that are broken.
+export function shouldExemptMethod (type) {
+    try {
+        const errorLines = new Error().stack.split('\n')
+        const errorFiles = new Set()
+        // Should cater for Chrome and Firefox stacks, we only care about https? resources.
+        const lineTest = /(\()?(http[^)]+):[0-9]+:[0-9]+(\))?/
+        for (const line of errorLines) {
+            const res = line.match(lineTest)
+            if (res) {
+                const path = res[2]
+                // checked already
+                if (errorFiles.has(path)) {
+                    continue
+                }
+                if (shouldExemptUrl(type, path)) {
+                    return true
+                }
+                errorFiles.add(res[2])
+            }
+        }
+    } catch (e) {
+        // Fall through
+    }
+    return false
+}
+
+// Iterate through the key, passing an item index and a byte to be modified
+export function iterateDataKey (key, callback) {
+    let item = key.charCodeAt(0)
+    for (const i in key) {
+        let byte = key.charCodeAt(i)
+        for (let j = 8; j >= 0; j--) {
+            callback(item, byte)
+
+            // find next item to perturb
+            item = nextRandom(item)
+
+            // Right shift as we use the least significant bit of it
+            byte = byte >> 1
+        }
+    }
+}
+
+export function isFeatureBroken (args, feature) {
+    return args.site.brokenFeatures.includes(feature)
+}
+
+// TODO make rollup aware of this so it can tree shake
+const mozProxies = 'wrappedJSObject' in window
+
+export class DDGProxy {
+    constructor (objectScope, property, proxyObject) {
+        if (mozProxies) {
+            this._native = objectScope.wrappedJSObject[property]
```

Comment:
This should be the only real functional change assuming everything was implemented correctly. We use wrappedJSObjects for firefox and load it into the content script rather than an injected content script into the page.